### PR TITLE
Feat/sort box/multi sorting

### PR DESCRIPTION
**Pull request by Leptopoda**
based on #590

I applied the requested changes from #585 and also some Dart3 cleanups.
Is the documentation enough?

### CODE REVIEW

**Comment by provokateurin** [comment]
Parts of https://github.com/nextcloud/neon/pull/590 also sneaked in here, please remove :)

**Comment by Leptopoda** [comment]
That's intnetional as I'll just drop the commit once the other one is merged.

To lazy to resolve conflicts twice :)
Review on the rest should be unaffected.

**Comment by provokateurin** [comment]
Ah right, I already approved the other one so please go ahead and merge&rebase.

**Comment by Leptopoda** [comment]
@provokateurin ping :)

File: packages/sort_box/lib/sort_box.dart [hunk 1]
```diff
@@ -1,58 +1,81 @@
-// ignore_for_file: public_member_api_docs
-
+/// Signature of a function returning a [Comparable].
 typedef ComparableGetter<T> = Comparable Function(T);
 
+/// Sorting Box to sort [List]s on multiple properties.
 class SortBox<T extends Enum, R> {
+  /// Constrcuts a new SortBox.
```
Comment (provokateurin) [inline packages/sort_box/lib/sort_box.dart:6]:
```suggestion
  /// Constructs a new SortBox.
```

File: packages/sort_box/lib/sort_box.dart [hunk 2]
```diff
@@ -1,58 +1,81 @@
...
-    this._secondaryBoxes,
+    this._boxes,
   );
 
+  /// A mapping of all values [T] to thier [ComparableGetter].
   final Map<T, ComparableGetter<R>> _properties;
-  final Map<T, Box<T>> _secondaryBoxes;
 
-  List<R> sort(final List<R> input, final Box<T> box) {
+  /// A mapping of values [T] to thier *Boxes*.
```
Comment (provokateurin) [inline packages/sort_box/lib/sort_box.dart:17]:
```suggestion
  /// A mapping of values [T] to their *Boxes*.
```

File: packages/sort_box/lib/sort_box.dart [hunk 3]
```diff
@@ -1,58 +1,81 @@
...
+  ///
+  /// The Boxes are applied if two elements are considered equal regarding thier property [T].
```
Comment (provokateurin) [inline packages/sort_box/lib/sort_box.dart:19]:
```suggestion
  /// The Boxes are applied if two elements are considered equal regarding their property [T].
```

File: packages/sort_box/lib/sort_box.dart [hunk 4]
```diff
@@ -1,58 +1,81 @@
...
+  final Map<T, Set<(T property, SortBoxOrder order)>> _boxes;
+
+  /// Sorts the [input] list according to thier [box].
```
Comment (provokateurin) [inline packages/sort_box/lib/sort_box.dart:22]:
```suggestion
  /// Sorts the [input] list according to their [box].
```

**Review by provokateurin — CHANGES_REQUESTED** [review]

File: packages/sort_box/lib/sort_box.dart [hunk 5]
```diff
@@ -1,58 +1,81 @@
...
+  /// Constructs a new SortBox.
+  ///
+  /// A *Box* is a record of a property and how to order it.
   SortBox(
     this._properties,
-    this._secondaryBoxes,
+    this._boxes,
   );
 
+  /// A mapping of all values [T] to thier [ComparableGetter].
```
Comment (provokateurin) [inline packages/sort_box/lib/sort_box.dart:14]:
```suggestion
  /// A mapping of all values [T] to their [ComparableGetter].
```

**Review by provokateurin — APPROVED** [review]
One last nit pick

File: packages/sort_box/lib/sort_box.dart (hunk 5 above)
Comment (Leptopoda) [reply packages/sort_box/lib/sort_box.dart:14]: Yeah lots of copy :spaghetti:

**Pull request merged**